dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ThreadCreation.java

```diff
@@ -0,0 +1,376 @@
+import com.dat3m.dartagnan.program.event.core.utils.RegWriter;
+import com.dat3m.dartagnan.program.event.functions.AbortIf;
+import com.dat3m.dartagnan.program.event.functions.DirectFunctionCall;
+import com.dat3m.dartagnan.program.event.functions.DirectValueFunctionCall;
+import com.dat3m.dartagnan.program.event.functions.Return;
+import com.dat3m.dartagnan.program.event.lang.llvm.LlvmCmpXchg;
+import com.dat3m.dartagnan.program.memory.Memory;
+import com.dat3m.dartagnan.program.memory.MemoryObject;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.sosy_lab.common.configuration.Configuration;
+import org.sosy_lab.common.configuration.InvalidConfigurationException;
+import org.sosy_lab.common.configuration.Option;
+import org.sosy_lab.common.configuration.Options;
+
+import java.math.BigInteger;
+import java.util.*;
+import java.util.stream.Stream;
+
+import static com.dat3m.dartagnan.configuration.OptionNames.TARGET;
+import static com.dat3m.dartagnan.configuration.OptionNames.THREAD_CREATE_ALWAYS_SUCCEEDS;
+import static com.dat3m.dartagnan.program.event.EventFactory.*;
+
+/*
+ * This pass handles (reachable) pthread-related function calls.
+ * - each pthread_create call spawns a new Thread object.
+ * - pthread_join calls are lowered to appropriate synchronization primitives.
+ * - get_my_tid calls are replaced by constant tid values.
+ * Initially, a single thread from the "main" function is spawned.
+ * Then the pass works iteratively by picking a (newly created) thread and handling all its pthread calls.
+ *
+ * TODO:
+ *  (1) Make sure that non-deterministic expressions are recreated properly (avoid wrong sharing)
+ *  (2) Make this pass able to run after compilation.
+ *  (3) Make sure that metadata is copied correctly.
+ */
+@Options
+public class ThreadCreation implements ProgramProcessor {
+
+    private static final Logger logger = LogManager.getLogger(ThreadCreation.class);
+
+    @Option(name = TARGET,
+            description = "The target architecture to which the program shall be compiled to.",
+            secure = true,
+            toUppercase = true)
+    private Arch compilationTarget = Arch.C11;
+
+    @Option(name = THREAD_CREATE_ALWAYS_SUCCEEDS,
+            description = "Calling pthread_create is guaranteed to succeed.",
+            secure = true,
+            toUppercase = true)
+    private boolean forceStart = false;
+
+    private ThreadCreation() {}
+
+    public static ThreadCreation fromConfig(Configuration config) throws InvalidConfigurationException {
+        ThreadCreation creation = new ThreadCreation();
+        config.inject(creation);
+        return creation;
+    }
+
+    @Override
+    public void run(Program program) {
+        if (program.getFormat().equals(Program.SourceLanguage.LITMUS)) {
+            return;
+        }
+        // TODO: test code
```

Comment:
Is this still test code?

Reply:
I could have sworn that I removed this line already... it's dead code now.